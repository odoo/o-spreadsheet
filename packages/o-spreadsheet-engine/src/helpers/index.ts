export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
