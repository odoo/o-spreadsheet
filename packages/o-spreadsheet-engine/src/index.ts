export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
