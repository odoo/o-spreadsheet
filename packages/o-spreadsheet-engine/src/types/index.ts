export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
