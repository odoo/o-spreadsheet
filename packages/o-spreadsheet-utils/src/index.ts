export * from "./translation";
export * from "./uuid";
