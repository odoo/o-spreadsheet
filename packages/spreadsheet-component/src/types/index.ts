/**
 * State
 *
 * This file defines the basic types involved in maintaining the running state
 * of a o-spreadsheet.
 *
 * The most important exported values are:
 * - interface GridState: the internal type of the state managed by the model
 */

import { Chart } from "chart.js";

export * from "./autofill";
export * from "./cells";
export * from "./chart/chart";
export * from "./clipboard";
export * from "./collaborative/revisions";
export * from "./collaborative/session";
export * from "./commands";
export * from "./conditional_formatting";
export * from "./currency";
export * from "./data_validation";
export * from "./env";
export * from "./figure";
export * from "./format";
export * from "./functions";
export * from "./getters";
export * from "./history";
export * from "./locale";
export * from "./misc";
export * from "./pivot";
export * from "./pivot_runtime";
export * from "./range";
export * from "./rendering";
export * from "./table";
export * from "./workbook_data";

declare global {
  interface Window {
    Chart: typeof Chart;
  }
}
