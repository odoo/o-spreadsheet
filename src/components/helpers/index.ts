export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
