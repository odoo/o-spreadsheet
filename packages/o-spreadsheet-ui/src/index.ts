import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}