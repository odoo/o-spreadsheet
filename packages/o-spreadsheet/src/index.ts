import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };