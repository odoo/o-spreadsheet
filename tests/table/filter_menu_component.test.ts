import { Model } from "../../src";
import { UID } from "../../src/types";
import {
  createDynamicTable,
  createTable,
  hideRows,
  setCellContent,
  setFormat,
  updateFilter,
} from "../test_helpers/commands_helpers";
import {
  focusAndKeyDown,
  keyDown,
  setInputValueAndTrigger,
  simulateClick,
} from "../test_helpers/dom_helper";
import { getCellsObject, mountSpreadsheet, nextTick } from "../test_helpers/helpers";

async function openFilterMenu() {
  await simulateClick(".o-filter-icon");
}

describe("Filter menu component", () => {
  let fixture: HTMLElement;
  let model: Model;
  let sheetId: UID;

  function getFilterMenuValues() {
    const values: { value: string; isChecked: boolean }[] = [];
    const filterValueEls = fixture.querySelectorAll(".o-filter-menu-value");
    for (const filterValue of filterValueEls) {
      const isChecked = !!filterValue.querySelector(".o-filter-menu-value-checked span")
        ?.textContent;
      const value = filterValue.querySelector(".o-filter-menu-value-text")!.textContent!;
      values.push({ value, isChecked });
    }
    return values;
  }

  beforeEach(async () => {
    ({ model, fixture } = await mountSpreadsheet());
    sheetId = model.getters.getActiveSheetId();
  });

  describe("Filter Tests", () => {
    beforeEach(async () => {
      createTable(model, "A1:A5");
      setCellContent(model, "A1", "header");
      setCellContent(model, "A2", "1");
      setCellContent(model, "A3", "1");
      setCellContent(model, "A4", "2");

      createTable(model, "B1:B4");
      setCellContent(model, "B2", "B2");
      setCellContent(model, "B3", "B3");
      setCellContent(model, "B4", "B4");
      await nextTick();
    });

    test("Filter menu is correctly rendered", async () => {
      await openFilterMenu();
      expect(fixture.querySelector(".o-filter-menu")).toMatchSnapshot();
    });

    test("Duplicates values are not displayed twice", async () => {
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "2"]);
    });

    test("Opening the filter menu of another filter update the values", async () => {
      const filterIcons = fixture.querySelectorAll(".o-filter-icon");
      await simulateClick(filterIcons[0]);
      let values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "2"]);

      await simulateClick(filterIcons[1]);
      values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["B2", "B3", "B4"]);
    });

    test("Uppercase/Lowercase are not displayed twice. Only the first occurrence is displayed", async () => {
      setCellContent(model, "A2", "PercevaL");
      setCellContent(model, "A3", "perceval");
      setCellContent(model, "A4", "gauviN");
      setCellContent(model, "A5", "Gauvin");

      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["gauviN", "PercevaL"]);
    });

    test("Displayed values are sorted using natural sorting", async () => {
      // Sort order should be 2,10,a,B and not 10,2,B,a (the default string comparison)
      setCellContent(model, "A2", "10");
      setCellContent(model, "A3", "2");
      setCellContent(model, "A4", "B");
      setCellContent(model, "A5", "a");

      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["2", "10", "a", "B"]);
    });

    test("We display the formated value of the cells", async () => {
      setFormat(model, "A4", "m/d/yyyy");
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1", "1/1/1900"]);
    });

    test("Values are checked depending on the filter state", async () => {
      updateFilter(model, "A1", ["1"]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: false },
        { value: "2", isChecked: true },
      ]);
    });

    test("Hidden values are not displayed", async () => {
      hideRows(model, [3]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "1"]);
    });

    test("Values hidden by another filter are not displayed", async () => {
      updateFilter(model, "B1", ["B2", "B3"]);
      await openFilterMenu();
      const values = getFilterMenuValues();
      expect(values.map((val) => val.value)).toEqual(["(Blanks)", "2"]);
    });

    test("Can hover mouse to select items", async () => {
      await openFilterMenu();
      const listItems = fixture.querySelectorAll(".o-filter-menu-value");
      expect(listItems[0].classList.contains("selected")).toBeFalsy();

      listItems[0].dispatchEvent(new Event("pointermove", { bubbles: true }));
      await nextTick();
      expect(listItems[0].classList.contains("selected")).toBeTruthy();
    });

    test("Clicking on values check and uncheck them", async () => {
      await openFilterMenu();
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: true });
      await simulateClick(".o-filter-menu-value");
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: false });
      await simulateClick(".o-filter-menu-value");
      expect(getFilterMenuValues()[0]).toEqual({ value: "(Blanks)", isChecked: true });
    });

    test("Confirm button updates the filter with formatted cell value", async () => {
      setFormat(model, "A4", "m/d/yyyy");
      expect(model.getters.getFilterHiddenValues({ sheetId, col: 0, row: 0 })).toEqual([]);
      await openFilterMenu();
      await simulateClick(".o-filter-menu-value:nth-of-type(2)");
      await simulateClick(".o-filter-menu-value:nth-of-type(3)");
      await simulateClick(".o-filter-menu-button-primary");
      expect(model.getters.getFilterHiddenValues({ sheetId, col: 0, row: 0 })).toEqual([
        "1",
        "1/1/1900",
      ]);
    });

    test("Cancel button don't save the changes", async () => {
      expect(model.getters.getFilterHiddenValues({ sheetId, col: 0, row: 0 })).toEqual([]);
      await openFilterMenu();
      await simulateClick(".o-filter-menu-value:nth-of-type(1)");
      await simulateClick(".o-filter-menu-value:nth-of-type(2)");
      await simulateClick(".o-filter-menu-button-cancel");
      expect(model.getters.getFilterHiddenValues({ sheetId, col: 0, row: 0 })).toEqual([]);
    });

    test("Can clear and select all", async () => {
      await openFilterMenu();
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: true },
        { value: "2", isChecked: true },
      ]);
      await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: false },
        { value: "1", isChecked: false },
        { value: "2", isChecked: false },
      ]);
      await simulateClick(".o-filter-menu-action-text:nth-of-type(1)");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: true },
        { value: "2", isChecked: true },
      ]);
    });

    test("Clear all work on the displayed values", async () => {
      await openFilterMenu();
      await setInputValueAndTrigger(".o-filter-menu input", "1");
      await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
      await setInputValueAndTrigger(".o-filter-menu input", "");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: true },
        { value: "1", isChecked: false },
        { value: "2", isChecked: true },
      ]);
    });

    test("Select all work on the displayed values", async () => {
      await openFilterMenu();
      await simulateClick(".o-filter-menu-action-text:nth-of-type(2)");
      await setInputValueAndTrigger(".o-filter-menu input", "1");
      await simulateClick(".o-filter-menu-action-text:nth-of-type(1)");
      await setInputValueAndTrigger(".o-filter-menu input", "");
      expect(getFilterMenuValues()).toEqual([
        { value: "(Blanks)", isChecked: false },
        { value: "1", isChecked: true },
        { value: "2", isChecked: false },
      ]);
    });

    test("Hitting esc key correctly closes the filter menu", async () => {
      await openFilterMenu();
      expect(fixture.querySelectorAll(".o-filter-menu")).toHaveLength(1);
      await keyDown({ key: "Escape" });
      expect(fixture.querySelectorAll(".o-filter-menu")).toHaveLength(0);
    });

    describe("Search bar", () => {
      test("Can filter values with the search bar", async () => {
        await openFilterMenu();
        await setInputValueAndTrigger(".o-filter-menu input", "1");
        expect(getFilterMenuValues().map((v) => v.value)).toEqual(["1"]);
      });

      test("Clicking on values focus the search input", async () => {
        await openFilterMenu();
        const searchInput = fixture.querySelector(".o-filter-menu input");
        expect(document.activeElement).not.toBe(searchInput);
        await simulateClick(".o-filter-menu-value");
        expect(document.activeElement).toBe(searchInput);
      });

      test("Search bar uses fuzzy search", async () => {
        setCellContent(model, "A2", "Florida");
        setCellContent(model, "A3", "Alaska");
        setCellContent(model, "A4", "Texas");
        setCellContent(model, "A5", "Illinois");

        await openFilterMenu();
        await setInputValueAndTrigger(".o-filter-menu input", "lo");
        expect(getFilterMenuValues().map((v) => v.value)).toEqual(["Florida", "Illinois"]);
      });

      test("Can use up/down arrow keys to select items", async () => {
        await openFilterMenu();

        await focusAndKeyDown(".o-filter-menu input", { key: "ArrowDown" });
        const listItems = fixture.querySelectorAll(".o-filter-menu-value");
        expect(listItems[0].classList.contains("selected")).toBeTruthy();

        await keyDown({ key: "ArrowDown" });
        expect(listItems[1].classList.contains("selected")).toBeTruthy();

        await keyDown({ key: "ArrowUp" });
        expect(listItems[0].classList.contains("selected")).toBeTruthy();
      });

      test("Can press enter to select/unselect items", async () => {
        await openFilterMenu();

        await focusAndKeyDown(".o-filter-menu input", { key: "ArrowDown" });
        await keyDown({ key: "Enter" });
        expect(getFilterMenuValues()[0].isChecked).toBeFalsy();

        await keyDown({ key: "Enter" });
        expect(getFilterMenuValues()[0].isChecked).toBeTruthy();
      });
    });
  });

  test("Sort filter", async () => {
    createTable(model, "A10:B15");
    setCellContent(model, "A10", "header");
    setCellContent(model, "A11", "olà");
    setCellContent(model, "A12", "1");
    setCellContent(model, "A13", "-1");
    setCellContent(model, "A14", "2");

    setCellContent(model, "B10", "header");
    setCellContent(model, "B11", "");
    setCellContent(model, "B12", "ab");
    setCellContent(model, "B13", "ba");
    setCellContent(model, "B14", "ca");
    await nextTick();

    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(1)");
    expect(getCellsObject(model, sheetId)).toMatchObject({
      A10: { content: "header" },
      A11: { content: "-1" },
      // A12 : empty cell, we give it a value of 0
      A13: { content: "1" },
      A14: { content: "2" },
      A15: { content: "olà" },
      B10: { content: "header" },
      B11: { content: "ba" },
      B13: { content: "ab" },
      B14: { content: "ca" },
    });

    await openFilterMenu();
    await simulateClick(".o-filter-menu-item:nth-of-type(2)");
    expect(getCellsObject(model, sheetId)).toMatchObject({
      A10: { content: "header" },
      A11: { content: "olà" },
      A12: { content: "2" },
      A13: { content: "1" },
      A15: { content: "-1" },
      B10: { content: "header" },
      B12: { content: "ca" },
      B13: { content: "ab" },
      B15: { content: "ba" },
    });
  });

  test("cannot sort filter table in readonly mode", async () => {
    createTable(model, "A10:B15");
    await nextTick();
    await openFilterMenu();
    expect(
      [...fixture.querySelectorAll(".o-filter-menu-item")].map((el) => el.textContent?.trim())
    ).toEqual(["Sort ascending (A ⟶ Z)", "Sort descending (Z ⟶ A)", "✓(Blanks)"]);
    model.updateMode("readonly");
    await nextTick();
    expect(
      [...fixture.querySelectorAll(".o-filter-menu-item")].map((el) => el.textContent?.trim())
    ).toEqual(["✓(Blanks)"]);
  });

  test("cannot sort dynamic table", async () => {
    setCellContent(model, "A10", "=MUNIT(2)");
    createDynamicTable(model, "A10");
    await nextTick();
    await openFilterMenu();
    expect(
      [...fixture.querySelectorAll(".o-filter-menu-item")].map((el) => el.textContent?.trim())
    ).not.toContain("Sort ascending (A ⟶ Z)");
  });
});
