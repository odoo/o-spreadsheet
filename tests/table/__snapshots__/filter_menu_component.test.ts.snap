// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter menu component Filter Tests Filter menu is correctly rendered 1`] = `
<div
  class="o-filter-menu d-flex flex-column bg-white"
>
  <div>
    <div
      class="o-filter-menu-item my-2 mb-3"
    >
       Sort ascending (A ⟶ Z) 
    </div>
    <div
      class="o-filter-menu-item my-2"
    >
       Sort descending (Z ⟶ A) 
    </div>
  </div>
  <div
    class="o-separator"
  />
  
  <div
    class="o-filter-menu-actions d-flex"
  >
    <div
      class="o-button-link me-4"
    >
      Select all
    </div>
    <div
      class="o-button-link me-4"
    >
      Clear
    </div>
  </div>
  <div
    class="position-relative"
  >
    <input
      class="w-100 o-input my-2"
      placeholder="Search..."
      type="text"
    />
    <i
      class="o-search-icon position-absolute"
    >
      <div
        class="o-icon"
      >
        <i
          class="fa fa-search"
        />
      </div>
    </i>
  </div>
  <div
    class="o-filter-menu-list d-flex flex-column"
  >
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <!-- toString because t-set with a body creates a LazyValue instead of a string -->
      <label
        class="o-checkbox d-flex align-items-center p-2 w-100 pe-auto text-truncate"
        role="button"
      >
        <input
          class="me-2"
          name="(Blanks)"
          type="checkbox"
        />
        (Blanks)
        
      </label>
      
    </div>
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <!-- toString because t-set with a body creates a LazyValue instead of a string -->
      <label
        class="o-checkbox d-flex align-items-center p-2 w-100 pe-auto text-truncate"
        role="button"
      >
        <input
          class="me-2"
          name="1"
          type="checkbox"
        />
        1
        
      </label>
      
    </div>
    <div
      class="o-filter-menu-item o-filter-menu-value"
    >
      <!-- toString because t-set with a body creates a LazyValue instead of a string -->
      <label
        class="o-checkbox d-flex align-items-center p-2 w-100 pe-auto text-truncate"
        role="button"
      >
        <input
          class="me-2"
          name="2"
          type="checkbox"
        />
        2
        
      </label>
      
    </div>
    
    
    
  </div>
  <div
    class="o-filter-menu-buttons d-flex justify-content-end"
  >
    <button
      class="o-button o-filter-menu-cancel me-2"
    >
      Cancel
    </button>
    <button
      class="o-button primary o-filter-menu-confirm"
    >
      Confirm
    </button>
  </div>
</div>
`;
